tests/tests.rs

```diff
@@ -156,6 +157,122 @@ async fn test_multithread_read_write() -> Result<(), String> {
     Ok(())
 }
 
+
+#[tokio::test(flavor = "multi_thread", worker_threads = 8)]
+async fn test_multithread_read_write_exist_delete() -> Result<(), String> {
+    let now = Instant::now();
+    let path = common::init("multithread_read_write_exist_delete");
+    let storage = Arc::new(
+        common::create_custom_test_storage(&path, |builder| {
+            builder
+                .max_blob_size(3_000_000)
+                .set_deferred_index_dump_times(Duration::from_millis(750), Duration::from_millis(1500))
+        }).await?);
+    const THREADS: usize = 2;
+    const KEYS_PER_THREAD: usize = 5000;
+    const OPERATIONS_PER_THREAD: usize = 25000;
+
+    let data = vec![184u8; 2000];
+    let handles: FuturesUnordered<_> = (0..THREADS)
+        .into_iter()
+        .map(|thread_id| {
+            let st = storage.clone();
+            let data = data.clone();
+            let task = async move {
+                let mut rng = ::rand::rngs::StdRng::from_entropy();
+                let mut map = HashMap::new();
+                let min_key: u32 = (thread_id * KEYS_PER_THREAD) as u32;
+                let max_key: u32 = ((thread_id + 1) * KEYS_PER_THREAD) as u32;
+
+                for _ in 0..OPERATIONS_PER_THREAD {
+                    let key: u32 = rng.gen_range(min_key..max_key);
+                    match rng.gen::<usize>() % 10 {
+                        0 => {
+                            st.delete(KeyTest::new(key), false).await.expect("delete success");
+                            map.insert(key, usize::MAX);
+                        },
+                        1 | 2 | 3 => {
+                            let data_len = rng.gen::<usize>() % data.len();
+                            write_one(&st, key, &data[0..data_len], None).await.expect("write success");
+                            map.insert(key, data_len);
+                        },
+                        4 | 5 | 6 => {
+                            let exist_res = st.contains(KeyTest::new(key)).await.expect("exist success");
+                            match exist_res {
```

Comment:
I think this check can be extracted into separate function (with flag for checking size in `found`), so test will be more concise

Reply:
It is not that easy, due to different result types. `contains` returns `ReadResult<BlobRecordTimestamp>`, `read` returns `ReadResult<Bytes>`. 

Other than that, `assert` inside will also print the line number, by which it is easier to diagnose the cause. If all checks are wrapped in a function, then the line number will not tell you anything

Reply:
I've changed the test code to compare `ReadResult<>` in `assert_eq!`. This required adding several helper function: one to map `Option<>` into `ReadResult<>`, other to reset timestamp inside `ReadResult::Deleted()`